Flaky test: TestGetNar_NixServeUpstream/nar_served_with_zstd_transparent_encoding — narinfo returns 404 under parallel subtest load

## Observed failure

In CI (`ncps-postgres-tests`, and likely other cohort derivations), the subtest
`TestGetNar_NixServeUpstream/nar_served_with_zstd_transparent_encoding` in
`pkg/server/server_test.go` fails intermittently with:

```
--- FAIL: TestGetNar_NixServeUpstream (0.14s)
    --- FAIL: TestGetNar_NixServeUpstream/nar_served_with_zstd_transparent_encoding (0.06s)
        server_test.go:1527:
                   Error Trace: github.com/kalbasit/ncps/pkg/server/server_test.go:1527
                   Error:       Not equal:
                                expected: 200
                                  actual: 404
                   Test:        TestGetNar_NixServeUpstream/nar_served_with_zstd_transparent_encoding
```

The failure is at the **narinfo** fetch step (line 1527), before the NAR is even
requested.  The test is fetching
`GET /{Nar7.NarInfoHash}.narinfo` and receiving 404 instead of 200.

## Context

`TestGetNar_NixServeUpstream` tests proxying from a nix-serve style upstream
(`Compression: none`, no `Content-Encoding` support).  It has three parallel
subtests that all share the **same ncps server instance and cache**:

1. `narinfo has Compression:none and FileSize==NarSize`
2. `nar served as raw bytes without Accept-Encoding`
3. `nar served with zstd transparent encoding` ← fails

All three subtests start by fetching the same narinfo (`Nar7`, `Compression:
none`).  When they run truly in parallel (as they do in the Nix sandbox build),
all three hit the narinfo endpoint simultaneously.  The suspected race: ncps
begins fetching + storing the narinfo on the first request, and concurrent
requests racing through the same code path can intermittently produce a 404
while the first fetch/store is in flight.

## Relationship to previous issues

Issues #227, #615, and #806 (all closed under milestone *v0.8.5 – Support for
nix-serve as upstream*) addressed correctness bugs with `Compression: none`
narinfos.  This is a different, narrower problem: a **concurrency / timing
issue** that only surfaces under parallel test execution in the sandbox.

The test itself was added in #1047 ("test: add transparent zstd serving
coverage") and passes reliably on `main` when the build is cached; it fails
when the derivation runs fresh (cache miss) and all subtests execute
simultaneously without prior warm-up.

## Reproduction

The failure appears in the `ncps-postgres-tests` (and probably `ncps-s3-tests`
and `ncps-mysql-tests`) Nix CI derivations.  It does not reproduce reliably with
`task test` locally because local tests share a process and the SQLite path is
more forgiving of concurrent access.

## Expected behaviour

All three subtests should independently be able to fetch the narinfo and receive
HTTP 200, regardless of parallel execution order.

## Suggested investigation

- Check whether ncps's narinfo fetch-and-store path is protected against
  concurrent requests for the same hash (i.e. does it hold a per-hash lock
  before fetching from upstream, or can two goroutines both receive "not found"
  and both attempt the upstream fetch simultaneously?).
- The prior fix in #882 addressed base16 vs nix32 hash handling; a similar
  in-flight deduplication guard may be needed for the concurrent-fetch case.

--- a/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/.openspec.yaml
+++ b/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/.openspec.yaml
@@ -1,0 +1,2 @@
+schema: spec-driven
+created: 2026-05-26

--- a/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/design.md
+++ b/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/design.md
@@ -1,0 +1,71 @@
+## Context
+
+`pkg/cache/cache.go` has two layers of concurrency control for upstream fetches:
+
+1. **`coordinateDownload`** (line ~5318): takes a `narInfoJobKey` or `narJobKey`, checks `upstreamJobs`, and if no job is in-flight, acquires a distributed lock (`downloadLocker`), re-checks `hasAsset`, then starts a goroutine and registers in `upstreamJobs`. Concurrent callers for the same key piggyback onto the running job via `ds.start`/`ds.stored`/`ds.done` channels.
+
+2. **`getNarInfoFromDatabase`** (line ~4078): called by `GetNarInfo` after the narinfo is stored, to validate it. Contains a purge guard: if `HasNarInStore(narURL)` returns false AND `hasUpstreamJob(narURL.Hash)` returns false AND `isRemoteDownloadInProgress` returns false, it purges the narinfo from DB and returns `errNarInfoPurged`.
+
+The race is in `getNarInfoFromDatabase`'s purge guard. The sequence of checks is non-atomic:
+
+```
+// goroutine A (concurrent request that piggybacked)
+1. HasNarInStore(narURL) → false        // NAR rename not yet visible
+   [goroutine B's NAR download completes: os.Rename(tmp→narPath)]
+   [goroutine B's pullNarInfo deferred done() removes job from upstreamJobs, closes ds.done]
+2. hasUpstreamJob(narURL.Hash) → false  // job just removed
+3. isRemoteDownloadInProgress  → false  // lock released
+4. PURGE triggered: narinfo deleted from DB
+```
+
+`pullNarInfo` calls `prePullNar` (starts NAR download goroutine) BEFORE `storeInDatabase`, so the narinfo lands in DB while the NAR download may still be in flight. When `ds.done` fires, piggybacking goroutines call `getNarInfoFromDatabase` and can hit the TOCTOU window above, triggering a spurious purge.
+
+A subsequent concurrent call then finds narinfo absent from DB, returns `storage.ErrNotFound`, and the HTTP handler returns 404.
+
+## Goals / Non-Goals
+
+**Goals:**
+- Eliminate the TOCTOU race in `getNarInfoFromDatabase`'s purge guard so concurrent narinfo fetches for the same hash never produce spurious 404s.
+- Maintain the existing `coordinateDownload` guarantees for both narinfo and NAR jobs.
+- Keep the fix confined to the cache layer; no changes to HTTP handlers or test structure.
+
+**Non-Goals:**
+- Rewriting the broader `coordinateDownload` / `upstreamJobs` machinery.
+- Adding per-request timeouts or retry logic to `GetNarInfo`.
+- Fixing any other race conditions beyond the narinfo purge TOCTOU.
+
+## Decisions
+
+**Fix the purge guard race by extending the job window through NAR download completion.**
+
+`pullNarInfo` currently fires `ds.done` (via the deferred `done()`) immediately after narinfo is stored in DB — before the NAR download goroutine spawned by `prePullNar` finishes. The purge guard in `getNarInfoFromDatabase` then has a window where neither `hasUpstreamJob` nor `hasNarInStore` is true.
+
+The fix: the `narInfoJobKey` job must not be removed from `upstreamJobs` until the NAR download has also completed (i.e., until the NAR is visible in the store). Concretely, the `done()` deferred in `pullNarInfo` should wait for the NAR-job channel (`narJobKey`) to complete before closing `ds.done`.
+
+Alternatively, extend `getNarInfoFromDatabase`'s purge check to also check whether a NAR job for the same hash is in-flight:
+
+```go
+// In getNarInfoFromDatabase, before purging:
+if hasUpstreamJob(narURL.Hash) || hasNarUpstreamJob(narURL.Hash) || isRemoteDownloadInProgress {
+    // don't purge
+}
+```
+
+This second approach is simpler and less invasive: it adds a check for the NAR job key (`"download:nar:" + hash`) alongside the existing narinfo job key check. If the NAR download is still running (job registered in `upstreamJobs`), the purge is skipped — the purge guard will be re-evaluated on the next read (the caller retries via the existing retry path).
+
+**Chosen approach: extend `getNarInfoFromDatabase` purge guard to check NAR job in-flight.**
+
+This is the minimal, targeted change:
+- One new helper or extend `hasUpstreamJob` to check both `narInfoJobKey` and `narJobKey`.
+- No changes to `pullNarInfo`, `coordinateDownload`, or channel signaling.
+- The fix is self-contained within the purge guard block.
+
+`hasUpstreamJob(hash)` currently checks `narJobKey(hash)` only. We will also check `narInfoJobKey(hash)` — or more precisely, `hasUpstreamJob` should be extended (or a new `hasAnyUpstreamJob(hash)` added) that returns true if EITHER key is registered.
+
+## Risks / Trade-offs
+
+- **Missed window shrinks but doesn't disappear**: the NAR job exists from when `coordinateDownload` registers it until its `done()` deferred fires. If the NAR download finishes and the job is removed between `HasNarInStore` and `hasAnyUpstreamJob`, the same TOCTOU window exists. However, `os.Rename` atomicity ensures that once `HasNarInStore` returns false, the rename hasn't happened, which means the NAR goroutine hasn't reached its final step. The job is still in `upstreamJobs` at this point. So checking both keys closes the gap.
+
+- **`isRemoteDownloadInProgress` still covers distributed scenarios**: the distributed lock path (`TryLock`) already handles the multi-instance case. The fix only tightens the single-instance race.
+
+- **Retry behavior**: when purge is skipped, `getNarInfoFromDatabase` still needs to return something useful to `GetNarInfo`. The existing error path (`errNarInfoPurged`) causes `GetNarInfo` to retry via `prePullNarInfo`. Since we're not purging, we can return the narinfo row directly (no change needed; purge is simply not triggered).

--- a/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/design.md
+++ b/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/design.md
@@ -8,7 +8,7 @@
 
 The race is in `getNarInfoFromDatabase`'s purge guard. The sequence of checks is non-atomic:
 
-```
+```text
 // goroutine A (concurrent request that piggybacked)
 1. HasNarInStore(narURL) → false        // NAR rename not yet visible
    [goroutine B's NAR download completes: os.Rename(tmp→narPath)]

--- a/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/proposal.md
+++ b/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/proposal.md
@@ -1,0 +1,27 @@
+## Why
+
+`TestGetNar_NixServeUpstream` has three parallel subtests that all fetch the same narinfo hash concurrently; when ncps's narinfo fetch-and-store path is not protected against concurrent requests for the same hash, competing goroutines can each receive "not found" and race to fetch from upstream — causing transient 404s in the subtests that lose the race. This manifests as a CI-only flake in the Nix sandbox derivations (`ncps-postgres-tests`, `ncps-s3-tests`, `ncps-mysql-tests`) where all subtests start simultaneously with a cold cache.
+
+## What Changes
+
+- Add a per-hash in-flight deduplication guard (e.g. `golang.org/x/sync/singleflight`) to the narinfo fetch-and-store path so concurrent requests for the same hash coalesce onto a single upstream fetch.
+- The test itself (`TestGetNar_NixServeUpstream`) requires no structural change — the fix makes the server behave correctly under parallel load.
+
+## Capabilities
+
+### New Capabilities
+
+- `narinfo-concurrent-fetch`: Coalescing / deduplication of concurrent upstream narinfo fetches for the same hash, ensuring at most one in-flight fetch per hash at any given time.
+
+### Modified Capabilities
+
+_(none — no existing spec-level requirements change)_
+
+## Impact
+
+- **`pkg/cache/`**: Core narinfo fetch-and-store logic gains a `singleflight.Group` (or equivalent) guard around the upstream fetch + local store sequence.
+- **`pkg/server/`**: No handler changes; the fix is in the cache layer below the HTTP handler.
+- **Tests**: The existing parallel subtests in `pkg/server/server_test.go` (`TestGetNar_NixServeUpstream`) will pass reliably without modification.
+- **Memory**: Negligible — `singleflight.Group` holds one in-flight result entry per hash during the fetch window only.
+- **Latency**: No added latency for the common (cache-warm) case; coalesced concurrent requests share one upstream round-trip instead of each making their own.
+- **Dependencies**: `golang.org/x/sync/singleflight` is already in the Go ecosystem standard library extension; no new external dependencies required.

--- a/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/specs/narinfo-concurrent-fetch/spec.md
+++ b/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/specs/narinfo-concurrent-fetch/spec.md
@@ -1,0 +1,56 @@
+## ADDED Requirements
+
+### Requirement: Concurrent narinfo requests for the same hash must not produce 404 responses
+
+When multiple requests arrive concurrently for a narinfo hash that is being fetched from upstream for the first time, all requests must eventually receive the narinfo (or a valid error), and none must receive a 404 caused by a spurious purge of the narinfo from the database.
+
+#### Scenario: Two concurrent requests for the same narinfo hash — second request must not get 404
+
+**Given** a cold cache (narinfo not yet stored locally)
+**And** a valid upstream narinfo for hash `H` exists
+**When** two requests for hash `H` arrive concurrently such that the second request calls `getNarInfoFromDatabase` while the NAR download goroutine for `H` is still running
+**Then** both requests receive the narinfo successfully
+**And** neither request receives a 404 or `storage.ErrNotFound`
+
+#### Scenario: Three concurrent requests for the same narinfo hash — all must succeed
+
+**Given** a cold cache
+**And** a valid upstream narinfo for hash `H` (with `Compression: none` and NAR size > 100KB) exists
+**When** three requests for hash `H` arrive in parallel (e.g. from `t.Parallel()` subtests sharing the same server)
+**Then** all three requests receive the narinfo successfully with HTTP 200
+**And** the narinfo is stored exactly once in the database
+**And** no spurious purge of the narinfo occurs
+
+#### Scenario: Purge guard skips purge when NAR download job is in-flight
+
+**Given** narinfo for hash `H` is stored in the database
+**And** a NAR download job for hash `H` is currently registered in `upstreamJobs`
+**When** `getNarInfoFromDatabase` is called and `HasNarInStore` returns false
+**Then** `getNarInfoFromDatabase` does NOT purge the narinfo from the database
+**And** it returns the narinfo to the caller
+**And** no `errNarInfoPurged` is triggered
+
+#### Scenario: Purge guard still fires when NAR is genuinely missing and no job is in-flight
+
+**Given** narinfo for hash `H` is stored in the database
+**And** the NAR file for hash `H` is absent from the store
+**And** no upstream job (`narInfoJobKey` or `narJobKey`) for hash `H` is registered
+**And** no remote download is in progress for hash `H`
+**When** `getNarInfoFromDatabase` is called
+**Then** the narinfo IS purged from the database
+**And** `errNarInfoPurged` is returned (triggering a fresh upstream fetch on the next request)
+
+### Requirement: The fix must not affect the common (cache-warm) case latency
+
+**Given** narinfo for hash `H` is already stored in the database and NAR is already in the store
+**When** a request for hash `H` arrives
+**Then** `getNarInfoFromDatabase` returns the narinfo directly without any additional synchronization overhead
+**And** no upstream fetch is triggered
+
+### Requirement: The fix must not alter behavior for single-hash sequential requests
+
+**Given** narinfo for hash `H` is fetched by a single request (no concurrency)
+**When** the fetch completes (narinfo stored, NAR downloaded)
+**And** subsequent requests arrive after the job has been removed from `upstreamJobs`
+**Then** `getNarInfoFromDatabase` behaves identically to before the fix
+**And** `HasNarInStore` returns true (NAR is in store) so the purge guard is not triggered

--- a/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/tasks.md
+++ b/openspec/changes/archive/2026-05-28-fix-flaky-narinfo-concurrent-fetch/tasks.md
@@ -1,0 +1,21 @@
+## 1. Write failing test
+
+- [x] 1.1 Add `TestGetNarInfo_ConcurrentFetch` to `pkg/cache/cache_test.go`: launch N goroutines that call `GetNarInfo` for the same hash simultaneously against a cold cache with a slow upstream stub (add ~10ms delay), assert all return the narinfo with no error
+- [x] 1.2 Run `task test` and confirm the new test fails (reproduces the race) before any production changes
+
+## 2. Fix purge guard in getNarInfoFromDatabase
+
+- [x] 2.1 Extend `hasUpstreamJob` (or add `hasAnyUpstreamJobForHash`) in `pkg/cache/cache.go` to return true when EITHER `narInfoJobKey(hash)` OR `narJobKey(hash)` is registered in `upstreamJobs`
+- [x] 2.2 Update the purge guard block in `getNarInfoFromDatabase` to call the extended helper so that a NAR download in-flight also prevents the spurious purge
+- [x] 2.3 Run `task test` and confirm the new `TestGetNarInfo_ConcurrentFetch` now passes
+
+## 3. Verify existing tests still pass
+
+- [x] 3.1 Run `task test` for the full `pkg/cache` and `pkg/server` packages with `-race` and confirm `TestGetNar_NixServeUpstream` and all other narinfo tests pass cleanly
+- [x] 3.2 Run `task fmt` and `task lint` — fix any formatting or lint issues before proceeding
+
+## 4. Completion checks
+
+- [x] 4.1 Run `task fmt` — confirm exits 0 with no changes
+- [x] 4.2 Run `task lint` — confirm exits 0
+- [x] 4.3 Run `task test` — confirm all tests pass with race detector

--- a/openspec/specs/narinfo-concurrent-fetch/spec.md
+++ b/openspec/specs/narinfo-concurrent-fetch/spec.md
@@ -1,0 +1,65 @@
+# narinfo-concurrent-fetch
+
+## Purpose
+
+Defines the correctness requirements for serving narinfo requests when multiple
+concurrent requests arrive for the same hash before the upstream fetch has
+completed. The core invariant is that no concurrent request may receive a 404
+caused by the purge guard firing while a NAR download job is still in-flight.
+
+## Requirements
+
+### Requirement: Concurrent narinfo requests for the same hash must not produce 404 responses
+
+When multiple requests arrive concurrently for a narinfo hash that is being fetched from upstream for the first time, all requests must eventually receive the narinfo (or a valid error), and none must receive a 404 caused by a spurious purge of the narinfo from the database.
+
+#### Scenario: Two concurrent requests for the same narinfo hash — second request must not get 404
+
+**Given** a cold cache (narinfo not yet stored locally)
+**And** a valid upstream narinfo for hash `H` exists
+**When** two requests for hash `H` arrive concurrently such that the second request calls `getNarInfoFromDatabase` while the NAR download goroutine for `H` is still running
+**Then** both requests receive the narinfo successfully
+**And** neither request receives a 404 or `storage.ErrNotFound`
+
+#### Scenario: Three concurrent requests for the same narinfo hash — all must succeed
+
+**Given** a cold cache
+**And** a valid upstream narinfo for hash `H` (with `Compression: none` and NAR size > 100KB) exists
+**When** three requests for hash `H` arrive in parallel (e.g. from `t.Parallel()` subtests sharing the same server)
+**Then** all three requests receive the narinfo successfully with HTTP 200
+**And** the narinfo is stored exactly once in the database
+**And** no spurious purge of the narinfo occurs
+
+#### Scenario: Purge guard skips purge when NAR download job is in-flight
+
+**Given** narinfo for hash `H` is stored in the database
+**And** a NAR download job for hash `H` is currently registered in `upstreamJobs`
+**When** `getNarInfoFromDatabase` is called and `HasNarInStore` returns false
+**Then** `getNarInfoFromDatabase` does NOT purge the narinfo from the database
+**And** it returns the narinfo to the caller
+**And** no `errNarInfoPurged` is triggered
+
+#### Scenario: Purge guard still fires when NAR is genuinely missing and no job is in-flight
+
+**Given** narinfo for hash `H` is stored in the database
+**And** the NAR file for hash `H` is absent from the store
+**And** no upstream job (`narInfoJobKey` or `narJobKey`) for hash `H` is registered
+**And** no remote download is in progress for hash `H`
+**When** `getNarInfoFromDatabase` is called
+**Then** the narinfo IS purged from the database
+**And** `errNarInfoPurged` is returned (triggering a fresh upstream fetch on the next request)
+
+### Requirement: The fix must not affect the common (cache-warm) case latency
+
+**Given** narinfo for hash `H` is already stored in the database and NAR is already in the store
+**When** a request for hash `H` arrives
+**Then** `getNarInfoFromDatabase` returns the narinfo directly without any additional synchronization overhead
+**And** no upstream fetch is triggered
+
+### Requirement: The fix must not alter behavior for single-hash sequential requests
+
+**Given** narinfo for hash `H` is fetched by a single request (no concurrency)
+**When** the fetch completes (narinfo stored, NAR downloaded)
+**And** subsequent requests arrive after the job has been removed from `upstreamJobs`
+**Then** `getNarInfoFromDatabase` behaves identically to before the fix
+**And** `HasNarInStore` returns true (NAR is in store) so the purge guard is not triggered

--- a/pkg/cache/cache.go
+++ b/pkg/cache/cache.go
@@ -5281,8 +5281,9 @@ func (c *Cache) hasUpstreamJob(hash string) bool {
 	defer c.upstreamJobsMu.Unlock()
 
 	_, narJobExists := c.upstreamJobs[narJobKey(hash)]
+	_, narInfoJobExists := c.upstreamJobs[narInfoJobKey(hash)]
 
-	return narJobExists
+	return narJobExists || narInfoJobExists
 }
 
 func (c *Cache) isRemoteDownloadInProgress(ctx context.Context, hash string) bool {

--- a/pkg/cache/cache.go
+++ b/pkg/cache/cache.go
@@ -4124,6 +4124,23 @@ func (c *Cache) getNarInfoFromDatabase(ctx context.Context, hash string) (*narin
 
 	// Check if this narinfo should be purged
 	if !hasNar && !isBeingDownloadedLocally && !isBeingDownloadedRemotely { //nolint:nestif // deferred
+		// Double-check HasNarInStore to close the TOCTOU window: if the NAR download
+		// completed (os.Rename) between the initial HasNarInStore check and hasUpstreamJob
+		// returning false (job removed after os.Rename), the NAR is already on disk.
+		hasNar = c.HasNarInStore(ctx, *narURL)
+		if !hasNar {
+			var recheckErr error
+
+			hasNar, recheckErr = c.HasNarInChunks(ctx, *narURL)
+			if recheckErr != nil {
+				return nil, fmt.Errorf("failed to recheck if nar exists in chunks: %w", recheckErr)
+			}
+		}
+
+		if hasNar {
+			return ni, nil
+		}
+
 		// For CDC: if a nar_files record exists with total_chunks=0, CDC chunking is
 		// in progress in a background goroutine (the job was removed from upstreamJobs
 		// when pullNarIntoStore returned early, before chunking finished). Don't purge.
@@ -5281,9 +5298,8 @@ func (c *Cache) hasUpstreamJob(hash string) bool {
 	defer c.upstreamJobsMu.Unlock()
 
 	_, narJobExists := c.upstreamJobs[narJobKey(hash)]
-	_, narInfoJobExists := c.upstreamJobs[narInfoJobKey(hash)]
 
-	return narJobExists || narInfoJobExists
+	return narJobExists
 }
 
 func (c *Cache) isRemoteDownloadInProgress(ctx context.Context, hash string) bool {

--- a/pkg/cache/cache_test.go
+++ b/pkg/cache/cache_test.go
@@ -2909,6 +2909,69 @@ func TestCacheBackends(t *testing.T) {
 	}
 }
 
+func testGetNarInfoConcurrentColdCacheFetch(factory cacheFactory) func(*testing.T) {
+	return func(t *testing.T) {
+		t.Parallel()
+
+		ts := testdata.NewTestServer(t, 40)
+		t.Cleanup(ts.Close)
+
+		c, _, _, _, _, cleanup := factory(t)
+		t.Cleanup(cleanup)
+
+		uc, err := upstream.New(newContext(), testhelper.MustParseURL(t, ts.URL), &upstream.Options{
+			PublicKeys: testdata.PublicKeys(),
+		})
+		require.NoError(t, err)
+
+		c.AddUpstreamCaches(newContext(), uc)
+
+		// Wait for upstream caches to become available.
+		<-c.GetHealthChecker().Trigger()
+
+		const concurrency = 10
+
+		var (
+			wg      sync.WaitGroup
+			mu      sync.Mutex
+			results = make([]*narinfo.NarInfo, 0, concurrency)
+			errs    []error
+		)
+
+		// Launch concurrent requests for the same hash against a cold cache.
+		// Nar7 has Compression:none and a 113KB NAR body, which creates a window
+		// where the NAR download has not completed when piggybacking goroutines
+		// call getNarInfoFromDatabase — the scenario that triggers a spurious purge.
+		for range concurrency {
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
+
+				ni, err := c.GetNarInfo(context.Background(), testdata.Nar7.NarInfoHash)
+
+				mu.Lock()
+				defer mu.Unlock()
+
+				if err != nil {
+					errs = append(errs, err)
+				} else {
+					results = append(results, ni)
+				}
+			}()
+		}
+
+		wg.Wait()
+
+		require.Empty(t, errs, "all concurrent GetNarInfo requests should succeed without error")
+		require.Len(t, results, concurrency, "all goroutines should receive the narinfo")
+
+		for i, ni := range results {
+			require.NotNil(t, ni, "result %d should not be nil", i)
+		}
+	}
+}
+
 func runCacheTestSuite(t *testing.T, factory cacheFactory) {
 	t.Helper()
 
@@ -2945,6 +3008,7 @@ func runCacheTestSuite(t *testing.T, factory cacheFactory) {
 	t.Run("GetNarInfoRaceConditionDuringMigrationDeletion", testGetNarInfoRaceConditionDuringMigrationDeletion(factory))
 	t.Run("GetNarInfoRaceConditionBeforeHasNarInfo", testGetNarInfoRaceConditionBeforeHasNarInfo(factory))
 	t.Run("GetNarInfoRaceWithPutNarInfoDeterministic", testGetNarInfoRaceWithPutNarInfoDeterministic(factory))
+	t.Run("GetNarInfoConcurrentColdCacheFetch", testGetNarInfoConcurrentColdCacheFetch(factory))
 	t.Run("testNarInfoFileSizeFix", testNarInfoFileSizeFix(factory))
 	t.Run("testCheckAndFixNarInfo", testCheckAndFixNarInfo(factory))
 	t.Run("HasNarFileRecord", testHasNarFileRecord(factory))

--- a/pkg/cache/cache_test.go
+++ b/pkg/cache/cache_test.go
@@ -2948,7 +2948,7 @@ func testGetNarInfoConcurrentColdCacheFetch(factory cacheFactory) func(*testing.
 			go func() {
 				defer wg.Done()
 
-				ni, err := c.GetNarInfo(context.Background(), testdata.Nar7.NarInfoHash)
+				ni, err := c.GetNarInfo(newContext(), testdata.Nar7.NarInfoHash)
 
 				mu.Lock()
 				defer mu.Unlock()

